### PR TITLE
[develop-poc] The Inspector

### LINKED ISSUE #13: Marshalling Object
Need some type marshaling for I/O

### PR DESCRIPTION
**Pull request by leonkozlowski**
#13 

**NEW**
-Added `src/jiggy/inspector.py` which introduces `Inspector`.
-This will serve as I/O marshall for both `node.params` and `node.output`
-Custom exceptions for both of above

**IMPROVEMENTS**
-Use inspector in `Sequential Runner`

**DISCUSSION**
-Requesting feedback on implementation though, the Inspector is instantiated at the `Node` level, the reason behind this, is that the accesses for the `Node` object include both `params` and `outputs`, which in turn helps runtime IMO, reduce need for loops, etc.

**FIXES**
-Fixed `no_ext.yml` so `function.output` is no longer a list (we discussed before)
-Some `Task` fixes as well.

### CODE REVIEW

File: src/jiggy/inspector.py [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+class InspectorParamException(Exception):
+    """Inspector Exception for param type mismatch"""
+
+    pass
+
+
+class InspectorOutputException(Exception):
+    """Inspector Exception for output type mismatch"""
+
+    pass
```
Comment (mitchbregs) [inline src/jiggy/inspector.py]: I think it's a good time for us to consider implementing `errors`.  My idea of implementation would be something like this:

https://github.com/Yelp/yelp-python/blob/master/yelp/errors.py

For now, totally solid.

Comment (mitchbregs) [inline src/jiggy/inspector.py:18]: I think it's a good time for us to consider implementing `errors`.  My idea of implementation would be something like this:

https://github.com/Yelp/yelp-python/blob/master/yelp/errors.py

For now, totally solid.

File: src/jiggy/inspector.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+        """Constructor for Inspector."""
+        self._node = node
+
+    def __repr__(self):
+        """Representation of Inspector."""
+        return "<Inspector {}>".format(self._node)
+
+    def inspect_param(self, param: dict, received: Any):
+        """Inspector helper for input and type."""
+        if not isinstance(received, eval(param.get("type"))):
```
Comment (mitchbregs) [inline src/jiggy/inspector.py:34]: How is this working with `datetime` for now?  Should we have a custom `eval` for future type implementations? I see that we are passing `datetime.datetime`.  Should it be handled this way?

File: src/jiggy/runner.py [hunk 3]
```diff
@@ -82,21 +83,31 @@ def __cls_run(init_cls: getattr, arguments=None):
     def _execute(self, pkg: str, mdl: str, node: Node, inputs=None):
         """Recursive executor for finding *args and **kwargs."""
         arguments = []
+        inspector = Inspector(node=node)
```
Comment (mitchbregs) [inline src/jiggy/runner.py:86]: Agree with placement here, this will eventually serve our DAG `Manager`

File: src/jiggy/task.py [hunk 4]
```diff
@@ -35,15 +35,16 @@ def source(self) -> str:
         return function.get("source", None) if function else None
 
     @property
-    def params(self) -> Union[list, None]:
+    def params(self) -> list:
         """Assign `task.function.params` property."""
         function = self.function
         return function.get("params", []) if function else None
```
Comment (mitchbregs) [inline src/jiggy/task.py:41]: You're doing this here `return function.get("params", []) if function else None` but not in line 46 below: `return function.get("output", {})`.. Why?

**Review by mitchbregs — APPROVED** [review]

File: src/jiggy/inspector.py (hunk 2 above)
Comment (leonkozlowski) [reply src/jiggy/inspector.py:34]: requires `import datetime`
my thought is that the use should get granular if necessary.

There is an inherent difference between something like:
`datetime.datetime` vs. `datetime.datetime.date`

File: src/jiggy/task.py (hunk 4 above)
Comment (leonkozlowski) [reply src/jiggy/task.py:41]: oversight -- will add

File: src/jiggy/inspector.py (hunk 1 above)
Comment (leonkozlowski) [reply src/jiggy/inspector.py:18]: we should kick out all exceptions to something like: `JiggyError`

File: src/jiggy/inspector.py (hunk 2 above)
Comment (mitchbregs) [reply src/jiggy/inspector.py:34]: Cool, I think later on we'll want some sort of `Fields` class that handles this for us. At the end of the day, `eval` is shaky, but I totally agree.

Comment (leonkozlowski) [reply src/jiggy/inspector.py:34]: Agree - suggesting we look at [marshmallow](https://marshmallow.readthedocs.io/en/stable/) for this

**Comment by mitchbregs** [comment]
+1

**Pull request merged**